pkg/tests/ossm/ior_test.go

```diff
@@ -0,0 +1,322 @@
+
+			t.LogStep("Deploy SMCP v2.3")
+			setupV23SMCP(t, meshNamespace, meshName)
+
+			t.LogStep("Check whether IOR creates Routes for hosts specified in the Gateway")
+			createSimpleGateway(t)
+			checkSimpleGateway(t)
+
+			t.LogStepf("Record the Route before the upgrade")
+			before := getRoutes(t, meshNamespace)
+
+			t.LogStep("Upgrade SMCP to v2.4")
+			updateToV24SMCP(t, meshNamespace, meshName)
+
+			t.LogStep("Check the Routes existed afther the upgrade")
+			checkSimpleGateway(t)
+			after := getRoutes(t, meshNamespace)
+
+			t.LogStep("Check the Routes unchanged afther the upgrade")
+			if before[0].Metadata.ResourceVersion != after[0].Metadata.ResourceVersion {
+				t.Fatal("Expect the route to be unchanged, but it is changed after the upgrade")
+			} else {
+				t.LogSuccess("Got the same resourceVersion before and after the upgrade")
+			}
+		})
+
+		t.NewSubTest("check IOR does not delete routes after deleting Istio pod").Run(func(t test.TestHelper) {
+			total := 3
+			nsNames := []string{}
+			gateways := []string{}
+
+			t.Cleanup(func() {
+				if env.GetSMCPVersion().GreaterThanOrEqual(version.SMCP_2_4) {
+					removeIORCustomSetting(t, meshNamespace, meshName)
+				}
+
+				oc.DeleteNamespace(t, nsNames...)
+				oc.ApplyString(t, meshNamespace, GetSMMRTemplate())
+				oc.WaitSMMRReady(t, meshNamespace)
+			})
+
+			t.LogStep("Ensure the IOR enabled")
+			if env.GetSMCPVersion().GreaterThanOrEqual(version.SMCP_2_4) {
+				enableIOR(t, meshNamespace, meshName)
+			}
+
+			t.LogStepf("Create %d Gateways and they are in their own Namespace", total)
+			for i := 0; i < total; i++ {
+				ns := fmt.Sprintf("ns-%d", i)
+				nsNames = append(nsNames, ns)
+				gateways = append(gateways, generateGateway(fmt.Sprintf("gw-%d", i), ns, fmt.Sprintf("www-%d.test.ocp", i)))
+			}
+			oc.CreateNamespace(t, nsNames...)
+			oc.ApplyString(t, "", gateways...)
+
+			t.LogStepf("Update SMMR to include %d Namespaces", total)
+			oc.ApplyString(t, meshNamespace, fmt.Sprintf(`
+apiVersion: maistra.io/v1
+kind: ServiceMeshMemberRoll
+metadata:
+  name: default
+spec:
+  members:
+  - bookinfo
+  - foo
+  - bar
+  - legacy
+  - %s
+  `, strings.Join(nsNames, "\n  - ")))
+
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				routes := getRoutes(t, meshNamespace)
+				if len(routes) == total {
+					t.LogSuccessf("Found all %d Routes", total)
+				} else {
+					t.Fatalf("Expect to find %d Routes but found %d instead", total, len(routes))
+				}
+			})
+
+			before := getRoutes(t, meshNamespace)
+			detectRouteChanges := func() {
+				retry.UntilSuccess(t, func(t test.TestHelper) {
+					after := getRoutes(t, meshNamespace)
+
+					if len(after) != total {
+						t.Fatalf("Expect %d Routes, but got %d instead", total, len(after))
+					}
+
+					if fmt.Sprint(buildRouteMap(before)) != fmt.Sprint(buildRouteMap(after)) {
+						t.Fatalf("Expect %d Routes remain unchanged, but they changed\nBefore: %v\nAfter: %v", total, before, after)
```

Comment:
```
ior_test.go:199:    FATAL: Expect 3 Routes remain unchanged, but they changed
        Before: [{{837ee85c-85f9-4894-ba94-22937fd3398c ns-0-gw-0-b3de2ff7f5c56cf4 63680 map[maistra.io/gateway-name:gw-0 maistra.io/gateway-namespace:ns-0 maistra.io/gateway-resourceVersion:63542 maistra.io/generated-by:ior] map[maistra.io/original-host:www-0.test.ocp]}} {{453b7180-bac2-4bbf-8bf4-3a437ad60b65 ns-1-gw-1-3df72d3a81944086 63684 map[maistra.io/gateway-name:gw-1 maistra.io/gateway-namespace:ns-1 maistra.io/gateway-resourceVersion:63544 maistra.io/generated-by:ior] map[maistra.io/original-host:www-1.test.ocp]}} {{8061eb68-dbec-4710-9a32-1318b7e80ffd ns-2-gw-2-1968729e121634b3 63682 map[maistra.io/gateway-name:gw-2 maistra.io/gateway-namespace:ns-2 maistra.io/gateway-resourceVersion:63546 maistra.io/generated-by:ior] map[maistra.io/original-host:www-2.test.ocp]}}]
        After: [{{837ee85c-85f9-4894-ba94-22937fd3398c ns-0-gw-0-b3de2ff7f5c56cf4 64162 map[maistra.io/gateway-name:gw-0 maistra.io/gateway-namespace:ns-0 maistra.io/gateway-resourceVersion:63542 maistra.io/generated-by:ior] map[maistra.io/original-host:www-0.test.ocp]}} {{453b7180-bac2-4bbf-8bf4-3a437ad60b65 ns-1-gw-1-3df72d3a81944086 64163 map[maistra.io/gateway-name:gw-1 maistra.io/gateway-namespace:ns-1 maistra.io/gateway-resourceVersion:63544 maistra.io/generated-by:ior] map[maistra.io/original-host:www-1.test.ocp]}} {{8061eb68-dbec-4710-9a32-1318b7e80ffd ns-2-gw-2-1968729e121634b3 64164 map[maistra.io/gateway-name:gw-2 maistra.io/gateway-namespace:ns-2 maistra.io/gateway-resourceVersion:63546 maistra.io/generated-by:ior] map[maistra.io/original-host:www-2.test.ocp]}}]
```

Hard to see what the difference is.

Reply:
Used `util.Compare` to generate [diff](https://github.com/maistra/maistra-test-tool/pull/375/files#diff-76a741e45a0b32014a31d2b4419dd39a56249aeeec8c494d9b82090f19af10ffR204) syntax for failure log.

Reply:
Changed to use yaml output for the routes and do the diff.